vllm/model_executor/layers/quantized_linear/gptq.py

```diff
@@ -0,0 +1,185 @@
+    qweight: torch.Tensor,
+    qzeros: torch.Tensor,
+    scales: torch.Tensor,
+    g_idx: torch.Tensor,
+    shifter: torch.Tensor,
+) -> torch.Tensor:
+    """Matrix multiplication with GPTQ weights."""
+    # qw: [input_size, output_size]
+    qw = (qweight.unsqueeze(1) >> shifter.view(1, -1, 1)) & 0xf
+    qw = qw.flatten(start_dim=0, end_dim=1)
+
+    # qz: [input_size, output_size]
+    qz = (qzeros[g_idx].unsqueeze(2) >> shifter.view(1, 1, -1)) & 0xf
+    qz = qz + 1
+    qz = qz.flatten(start_dim=1, end_dim=2)
+
+    # qs: [input_size, output_size]
+    qs = scales[g_idx]
+    # w: [input_size, output_size]
+    w = qs * (qw - qz).to(qs.dtype)
+
+    # out: [batch_size, output_size]
+    out = torch.matmul(x, w)
+    return out
+
+
+class GPTQColumnParallelLinear(ColumnParallelLinear):
+
+    def create_weights(self, dtype: torch.dtype) -> None:
+        assert self.input_size % self.quant_config.pack_factor == 0
+        assert self.input_size % self.quant_config.group_size == 0
+        assert (self.output_size_per_partition %
+                self.quant_config.pack_factor == 0)
+
+        group_size = self.quant_config.group_size
+        if group_size == -1:
+            group_size = self.input_size
+
+        self.qweight = Parameter(
+            torch.empty(
+                self.input_size // self.quant_config.pack_factor,
+                self.output_size_per_partition,
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.qzeros = Parameter(
+            torch.empty(
+                self.input_size // group_size,
+                self.output_size_per_partition //
+                self.quant_config.pack_factor,
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.scales = Parameter(
+            torch.empty(
+                self.input_size // group_size,
+                self.output_size_per_partition,
+                device="cuda",
+                dtype=dtype,
+            ),
+            requires_grad=False,
+        )
+        # Initialize g_idx to be sequential.
+        # This is required because old GPTQ models may not have g_idx.
+        self.g_idx = Parameter(
+            torch.tensor(
+                [i // group_size for i in range(self.input_size)],
+                device="cuda",
+                dtype=torch.int32,
+            ),
+            requires_grad=False,
+        )
+        self.shifter = torch.tensor(
+            [0, 4, 8, 12, 16, 20, 24, 28],
+            device="cuda",
+            dtype=torch.int32,
+        )
+
+    def apply_weights(
+        self,
+        x: torch.Tensor,
+        bias: Optional[torch.Tensor],
+    ) -> torch.Tensor:
+        out_shape = x.shape[:-1] + (self.qweight.shape[-1], )
+        reshaped_x = x.reshape(-1, x.shape[-1])
+        num_tokens = x.shape[:-1].numel()
+        if num_tokens <= 32:
```

Comment:
Why do we have an `if` here? Is the CUDA kernel slower when `num_tokens > 32` or it's just the CUDA kernel will not work at all?

Reply:
Good point. Actually, the current GPTQ kernel is designed for batch size 1 and performs extremely poorly when the batch size is large, often taking 10+ minutes for the initial memory profiling. As a workaround, I implemented a simple PyTorch-based GPTQ matmul faster than the original kernel for large batch size. Still, the two implementations are very bad and probably much slower than the optimized implementations like exllama.

Reply:
Added a comment on this.